### ISSUE TITLE
autoOutdent functions take 'session', not 'document'

### ISSUE THREAD

**Issue by kevinushey** [ajaxorg/ace#2417]
The `autoOutdent` for (almost) all modes specify the prototype as:

    this.autoOutdent = function(state, doc, row) { ... }

However, `autoOutdent` is used within [editor.js](https://github.com/ajaxorg/ace/blob/master/lib/ace/editor.js#L1019) as:

    mode.autoOutdent(lineState, session, cursor.row);

Should the `mode` prototypes be amended to read

    this.autoOutdent = function(state, session, row) { ... }

?

**Comment by nightwing** [comment]
Yes, doc name remains from time when ace didn't have session, it can be changed to session, but autoOutdent itself needs some refactoring, it probably needs to be merged with behaviors.
